src/intrinsics/ecma262/Object.js

```diff
@@ -36,11 +37,170 @@ import {
+    if (to.hasStringOrSymbolProperties()) {
+      // preserve them in a snapshot and add the snapshot to the sources
+      delayedSources.push(to.getSnapshot({ removeProperties: true }));
+    }
+
+    if (frm instanceof ObjectValue && frm.mightBeHavocedObject()) {
+      // it's not safe to trust any of its values
+      delayedSources.push(frm);
+    } else if (frm_was_partial) {
+      if (frm instanceof AbstractObjectValue && frm.kind === "explicit conversion to object") {
+        // Make it implicit again since it is getting delayed into an Object.assign call.
+        delayedSources.push(frm.args[0]);
+      } else {
+        let frmSnapshot = frm.getSnapshot();
+        frm.temporalAlias = frmSnapshot;
+        frm.makePartial();
+        delayedSources.push(frmSnapshot);
+      }
+    }
+  }
+}
+
+function copyKeys(realm: Realm, keys, from, to): void {
+  // c. Repeat for each element nextKey of keys in List order,
+  for (let nextKey of keys) {
+    // i. Let desc be ? from.[[GetOwnProperty]](nextKey).
+    let desc = from.$GetOwnProperty(nextKey);
+
+    // ii. If desc is not undefined and desc.[[Enumerable]] is true, then
+    if (desc && desc.enumerable) {
+      Props.ThrowIfMightHaveBeenDeleted(desc.value);
+
+      // 1. Let propValue be ? Get(from, nextKey).
+      let propValue = Get(realm, from, nextKey);
+
+      // 2. Perform ? Set(to, nextKey, propValue, true).
+      Props.Set(realm, to, nextKey, propValue, true);
+    }
+  }
+}
+
+function applyObjectAssignSource(
+  realm: Realm,
+  nextSource: ObjectValue | AbstractObjectValue,
+  to: ObjectValue | AbstractObjectValue,
+  delayedSources: Array<Value>,
+  to_must_be_partial: boolean
+): boolean {
+  let keys, frm;
+
+  // a. If nextSource is undefined or null, let keys be a new empty List.
+  if (HasSomeCompatibleType(nextSource, NullValue, UndefinedValue)) {
+    return to_must_be_partial;
+  }
+
+  // b. Else,
+  // i. Let from be ToObject(nextSource).
+  frm = To.ToObject(realm, nextSource);
+
+  // ii. Let keys be ? from.[[OwnPropertyKeys]]().
+  let frm_was_partial = frm.isPartialObject();
+  if (frm_was_partial) {
+    if (!to.isSimpleObject() || !frm.isSimpleObject()) {
+      // If an object is not a simple object, it may have getters on it that can
+      // mutate any state as a result. We don't yet support this.
+      AbstractValue.reportIntrospectionError(nextSource);
+      throw new FatalError();
+    }
+
+    to_must_be_partial = true;
+    // Make this temporarily not partial
+    // so that we can call frm.$OwnPropertyKeys below.
+    frm.makeNotPartial();
+    try {
+      keys = frm.$OwnPropertyKeys();
+
+      if (to_must_be_partial) {
+        handleObjectAssignSnapshot(to, frm, frm_was_partial, delayedSources);
+      }
+    } finally {
+      frm.makePartial();
+    }
+  } else {
+    keys = frm.$OwnPropertyKeys();
+
+    if (to_must_be_partial) {
+      handleObjectAssignSnapshot(to, frm, frm_was_partial, delayedSources);
+    }
```

Comment:
This looks duplicated from the above code path, can we unify?
You can make `makePartial` and `makeNotPartial` calls conditional but leave the `try` / `finally` structure shared.

Reply:
I did that originally but it looked more complex. Unless you have an example that looks more DRY.

Reply:
This reads simpler to me but maybe I'm missing something

```js
  // ii. Let keys be ? from.[[OwnPropertyKeys]]().
  let frm_was_partial = frm.isPartialObject();
  if (frm_was_partial) {
    if (!to.isSimpleObject() || !frm.isSimpleObject()) {
      // If an object is not a simple object, it may have getters on it that can
      // mutate any state as a result. We don't yet support this.
      AbstractValue.reportIntrospectionError(nextSource);
      throw new FatalError();
    }
    to_must_be_partial = true;
    // Make this temporarily not partial
    // so that we can call frm.$OwnPropertyKeys below.
    frm.makeNotPartial();
  }

  try {
    keys = frm.$OwnPropertyKeys();
    if (to_must_be_partial) {
      handleObjectAssignSnapshot(to, frm, frm_was_partial, delayedSources);
    }
  } finally {
    if (frm_was_partial) {
      frm.makePartial();
    }
  }

```